h/static/scripts/base/controller.js

```diff
@@ -25,6 +25,23 @@ function findRefs(el) {
 }
 
 /**
+ * Dispatch an event to all event handlers of an element's controller registered
+ * using `Controller#on`
+ */
+function dispatchEvent(element, event) {
+  if (!element.controllers) {
+    return;
+  }
+  element.controllers.forEach(ctrl => {
+    ctrl._listeners.forEach(listener => {
```

Comment:
I don't think an array is the appropriate data structure here. Let's have `_listeners` be a mapping from event type to array of listeners, then you don't need to use `forEach` here.